Allow Hardware Intrinsics to work with the AltJit

tannergooding opened the pull request:
There were two issues blocking this:
1. The `*.PlatformNotSupported.cs` files weren't marking all methods as `[Intrinsic]` (outside `IsSupported`)
2. The VM was restricting the type handling from recursively marking methods in `[Intrinsic]` types for mismatched platforms

tannergooding commented:
This allows the following to "just work" with an AltJit (noting the use of `fadd`):
```asm
; Assembly listing for method Program:Main(System.String[])
; Emitting BLENDED_CODE for generic ARM64 CPU - Windows
; optimized code
; fp based frame
; partially interruptible
; No PGO data
; invoked as altjit
; Final local variable assignments
;
;* V00 arg0         [V00    ] (  0,  0   )     ref  ->  zero-ref    class-hnd single-def
;# V01 OutArgs      [V01    ] (  1,  1   )  lclBlk ( 0) [sp+00H]   "OutgoingArgSpace"
;  V02 tmp1         [V02,T00] (  3,  6   )     ref  ->   x0         class-hnd exact single-def "Single-def Box Helper"
;
; Lcl frame size = 0

G_M24006_IG01:              ;; offset=0000H
        A9BF7BFD          stp     fp, lr, [sp, #-0x10]!
        910003FD          mov     fp, sp
                                                ;; size=8 bbWeight=1    PerfScore 1.50
G_M24006_IG02:              ;; offset=0008H
        D29FBB00          movz    x0, #0xFDD8
        F2B70020          movk    x0, #0xB801 LSL #16
        F2CFFF20          movk    x0, #0x7FF9 LSL #32
        94000000          bl      CORINFO_HELP_NEWSFAST
        6F000410          mvni    v16.4s, #0
        4F000411          movi    v17.4s, #0
        4E31D610          fadd    v16.4s, v16.4s, v17.4s
        3C808010          str     q16, [x0, #0x08]
        D292F901          movz    x1, #0x97C8      // code for System.Console:WriteLine
        F2B70121          movk    x1, #0xB809 LSL #16
        F2CFFF21          movk    x1, #0x7FF9 LSL #32
        F9400021          ldr     x1, [x1]
        D63F0020          blr     x1
                                                ;; size=52 bbWeight=1    PerfScore 13.00
G_M24006_IG03:              ;; offset=003CH
        A8C17BFD          ldp     fp, lr, [sp], #0x10
        D65F03C0          ret     lr
                                                ;; size=8 bbWeight=1    PerfScore 2.00

; Total bytes of code 68, prolog size 8, PerfScore 23.30, instruction count 17, allocated bytes for code 68 (MethodHash=7b7ba239) for method Program:Main(System.String[])
; ============================================================
```

Prior to this we emitted the following instead (note the `PlatformNotSupportedException`):
```
; Assembly listing for method Program:Main(System.String[])
; Emitting BLENDED_CODE for generic ARM64 CPU - Windows
; optimized code
; fp based frame
; partially interruptible
; No PGO data
; 0 inlinees with PGO data; 1 single block inlinees; 0 inlinees without PGO data
; invoked as altjit
; Final local variable assignments
;
;* V00 arg0         [V00    ] (  0,  0   )     ref  ->  zero-ref    class-hnd single-def
;# V01 OutArgs      [V01    ] (  1,  1   )  lclBlk ( 0) [sp+00H]   "OutgoingArgSpace"
;  V02 tmp1         [V02,T00] (  3,  6   )     ref  ->  x19         class-hnd exact single-def "NewObj constructor temp"
;* V03 tmp2         [V03    ] (  0,  0   )     ref  ->  zero-ref    class-hnd exact "Single-def Box Helper"
;
; Lcl frame size = 8

G_M24006_IG01:              ;; offset=0000H
        A9BE7BFD          stp     fp, lr, [sp, #-0x20]!
        F9000FF3          str     x19, [sp, #0x18]
        910003FD          mov     fp, sp
                                                ;; size=12 bbWeight=1    PerfScore 2.50
G_M24006_IG02:              ;; offset=000CH
        D2822700          movz    x0, #0x1138
        F2BFBF20          movk    x0, #0xFDF9 LSL #16
        F2CFFF20          movk    x0, #0x7FF9 LSL #32
        94000000          bl      CORINFO_HELP_NEWSFAST
        AA0003F3          mov     x19, x0
        D287DA01          movz    x1, #0x3ED0      // code for System.PlatformNotSupportedException:.ctor
        F2BFBF01          movk    x1, #0xFDF8 LSL #16
        F2CFFF21          movk    x1, #0x7FF9 LSL #32
        F9400021          ldr     x1, [x1]
        D63F0020          blr     x1
        AA1303E0          mov     x0, x19
        94000000          bl      CORINFO_HELP_THROW
        D43E0000          brk_windows #0
                                                ;; size=52 bbWeight=1    PerfScore 11.00

; Total bytes of code 64, prolog size 12, PerfScore 19.90, instruction count 16, allocated bytes for code 64 (MethodHash=7b7ba239) for method Program:Main(System.String[])
; ============================================================
```

kasperk81 commented on src/coreclr/vm/methodtablebuilder.cpp:
```diff
@@ -1493,13 +1493,8 @@ MethodTableBuilder::BuildMethodTableThrowing(
-        if (hr == S_OK && (strcmp(nameSpace, "System.Runtime.Intrinsics.X86") == 0))
-#endif
+        // All the funtions in System.Runtime.Intrinsics are potentially hardware intrinsics.
+        if (hr == S_OK && strncmp(nameSpace, "System.Runtime.Intrinsics", 25) == 0)
```
nit
```c
const char* prefix = "System.Runtime.Intrinsics";
if (hr == S_OK && strncmp(nameSpace, prefix, strlen(prefix)) == 0)
```

tannergooding replied:
I don't believe we currently follow a similar pattern elsewhere.

There are also considerations that we have many wrappers for various CRT functions in the PAL layers and so `strlen` might not be constant folded as intended.

tannergooding commented:
@jkotas, any concerns with this relatively minor VM change (or the larger change in general)?

jkotas commented:
Matching intrinsics by name is relatively expensive. It sounds like we will try to do that much more often with this change. Is that right? How much JIT throughput (for intrinisic heavy code) is this going to cost?

We typically make the AltJit paths via slow paths. For example, we can assume that any corelib method is intrinsics and try to match it by name when running as cross-targeting AltJit.

tannergooding commented:
> Matching intrinsics by name is relatively expensive. It sounds like we will try to do that much more often with this change. Is that right? How much JIT throughput (for intrinisic heavy code) is this going to cost?

On x64 we'll start matching any call that looks like `System.Runtime.Intrinsics.Arm` and on Arm64 any that looks like `System.Runtime.Intrinsics.x64`.

However, would be able to directly recognize such cases as `NI_Throw_PlatformNotSupportedException` if desired and so can ultimately "save" cost on inlining.

Is there a way to explicitly trigger `superpmi-diffs`? That would give us some basic throughput metrics to see if this is currently a positive or negative impact.

> We typically make the AltJit paths via slow paths. For example, we can assume that any corelib method is intrinsics and try to match it by name when running as cross-targeting AltJit.

I could certainly do that, but I worry that would be significantly more expensive, especially considering the standard patterns most HWIntrinsic usage is expected to follow.

jkotas commented:
> would be able to directly recognize such cases as NI_Throw_PlatformNotSupportedException if desired and so can ultimately "save" cost on inlining.

I do not see where we would save here:

If we are compiling with optimizations on, we should be dead-code eliminating the unreachable blocks under IsSupported checks early enough before any of them hit inliner.

If we are compiling with optimizations off, we do not inline anything anyway.

jkotas commented:
> I could certainly do that, but I worry that would be significantly more expensive

It does not matter for altjit. cross-plat altjit is slow in other places, and also only like 99% fidelity.

tannergooding commented:
Makes sense, I'll see about updating it to work that way instead.

tannergooding committed:
Fixing a case where a *.PlatformNotSupported.cs type was marked Intrinsic

tannergooding committed:
Under mismatched VM (AltJit) allow any method to be checked as an intrinsic

tannergooding committed:
Simplify some logic in impIntrinsic and move the info.compMatchedVM check higher

tannergooding commented:
Adjusted to always check when `!info.compMatchedVM`.

`impIntrinsic` notably still had some logic from when we had `INTRINSIC` and `JIT INTRINSIC` concepts and therefore two flags. So I simplified the check in `impIntrinsic`.

tannergooding commented:
The 1 unmarked job did complete and complete successfully, it just never reported.
